File: api/internal/transport/carrier.go
// Carrier seams — the bring-your-own-carrier boundary (conventions doc §7).
//
// The library owns envelope codecs, framing, and lifecycle; the *carrier* (the
// byte/datagram transport) is injected. A host supplies QUIC, WebRTC, a platform
// media stack, or anything else by implementing one of these interfaces — without
// changing the library.
package transport

import (
	"encoding/binary"
	"errors"
	"io"
)

// ErrCarrier is a carrier-level transport error (a wrapped I/O failure or a
// protocol violation by the carrier).
type ErrCarrier string

func (e ErrCarrier) Error() string { return "carrier error: " + string(e) }

// FrameCarrier sends and receives one *delimited message* at a time. Used by
// CSIL-RPC and CSIL-Events. Built-in implementations frame with a 4-byte
// big-endian length prefix; a host may implement this over WebSocket binary
// frames, a WebTransport stream, etc.
type FrameCarrier interface {
	SendFrame(bytes []byte) error
	// RecvFrame receives the next frame, or nil at a clean end of stream.
	RecvFrame() ([]byte, error)
}

// DatagramCarrier sends and receives one self-contained datagram (each within the
// channel MTU), with no delivery or ordering guarantee. Used by CSIL-Datagrams.
// Built-in over UDP; a host plugs WebRTC unreliable channels, QUIC datagrams, etc.
type DatagramCarrier interface {
	SendDatagram(bytes []byte) error
	// RecvDatagram receives the next datagram, or nil if the carrier is closed.
	RecvDatagram() ([]byte, error)
}

// WriteLengthPrefixed writes a 4-byte big-endian length prefix followed by bytes
// (CSIL stream framing), enforcing the max-frame guard before writing.
func WriteLengthPrefixed(w io.Writer, b []byte, max int) error {
	if len(b) > max {
		return ErrFrameTooLarge{Got: len(b), Max: max}
	}
	var prefix [4]byte
	binary.BigEndian.PutUint32(prefix[:], uint32(len(b)))
	if _, err := w.Write(prefix[:]); err != nil {
		return ErrCarrier(err.Error())
	}
	if _, err := w.Write(b); err != nil {
		return ErrCarrier(err.Error())
	}
	return nil
}

// ReadLengthPrefixed reads one length-prefixed frame, enforcing the max-frame
// guard before allocating. It returns nil at a clean EOF before any byte of a frame.
func ReadLengthPrefixed(r io.Reader, max int) ([]byte, error) {
	var lenBuf [4]byte
	_, err := io.ReadFull(r, lenBuf[:])
	if err != nil {
		// A clean EOF before any frame byte is an orderly end of stream.
		if errors.Is(err, io.EOF) {
			return nil, nil
		}
		return nil, ErrCarrier(err.Error())
	}
	// Compare the prefix as an unsigned value before narrowing to int: on a 32-bit
	// platform a length >= 0x80000000 would become a negative int that slips past a
	// signed > max guard and then panics in make([]byte, n).
	length := binary.BigEndian.Uint32(lenBuf[:])
	if uint64(length) > uint64(max) {
		return nil, ErrFrameTooLarge{Got: int(length), Max: max}
	}
	buf := make([]byte, length)
	if _, err := io.ReadFull(r, buf); err != nil {
		return nil, ErrCarrier(err.Error())
	}
	return buf, nil
}

// StreamCarrier is a FrameCarrier over any io.ReadWriter byte stream (TCP, TLS,
// Unix socket), using the canonical 4-byte length-prefix framing.
type StreamCarrier struct {
	stream   io.ReadWriter
	maxFrame int
}

func NewStreamCarrier(stream io.ReadWriter) *StreamCarrier {
	return &StreamCarrier{stream: stream, maxFrame: MaxFrameDefault}
}

// NewStreamCarrierWithMaxFrame builds a carrier with a host-chosen max-frame
// limit. The limit is validated here rather than at the first frame, so a
// misconfigured carrier is a construction-time error the host can surface at
// startup.
func NewStreamCarrierWithMaxFrame(stream io.ReadWriter, maxFrame int) (*StreamCarrier, error) {
	if err := ValidateMaxFrame(maxFrame); err != nil {
		return nil, err
	}
	return &StreamCarrier{stream: stream, maxFrame: maxFrame}, nil
}

// MaxFrame reports the limit this carrier enforces in both directions.
func (c *StreamCarrier) MaxFrame() int { return c.maxFrame }

func (c *StreamCarrier) SendFrame(b []byte) error {
	return WriteLengthPrefixed(c.stream, b, c.maxFrame)
}

func (c *StreamCarrier) RecvFrame() ([]byte, error) {
	return ReadLengthPrefixed(c.stream, c.maxFrame)
}

// LoopbackFrameCarrier is an in-memory FrameCarrier backed by queues of frames —
// for tests and for driving the codec without a socket.
type LoopbackFrameCarrier struct {
	Outbound [][]byte
	Inbound  [][]byte
}

func NewLoopbackFrameCarrier() *LoopbackFrameCarrier { return &LoopbackFrameCarrier{} }

// PushInbound queues a frame that a subsequent RecvFrame will return.
func (c *LoopbackFrameCarrier) PushInbound(b []byte) {
	c.Inbound = append(c.Inbound, b)
}

// TakeOutbound takes the next frame that was sent via SendFrame, or nil if none.
func (c *LoopbackFrameCarrier) TakeOutbound() []byte {
	if len(c.Outbound) == 0 {
		return nil
	}
	f := c.Outbound[0]
	c.Outbound = c.Outbound[1:]
	return f
}

func (c *LoopbackFrameCarrier) SendFrame(b []byte) error {
	cp := make([]byte, len(b))
	copy(cp, b)
	c.Outbound = append(c.Outbound, cp)
	return nil
}

func (c *LoopbackFrameCarrier) RecvFrame() ([]byte, error) {
	if len(c.Inbound) == 0 {
		return nil, nil
	}
	f := c.Inbound[0]
	c.Inbound = c.Inbound[1:]
	return f, nil
}

// LoopbackDatagramCarrier is an in-memory DatagramCarrier — for tests and codec drives.
type LoopbackDatagramCarrier struct {
	Outbound [][]byte
	Inbound  [][]byte
}

func NewLoopbackDatagramCarrier() *LoopbackDatagramCarrier { return &LoopbackDatagramCarrier{} }

func (c *LoopbackDatagramCarrier) PushInbound(b []byte) {
	c.Inbound = append(c.Inbound, b)
}

func (c *LoopbackDatagramCarrier) TakeOutbound() []byte {
	if len(c.Outbound) == 0 {
		return nil
	}
	f := c.Outbound[0]
	c.Outbound = c.Outbound[1:]
	return f
}

func (c *LoopbackDatagramCarrier) SendDatagram(b []byte) error {
	cp := make([]byte, len(b))
	copy(cp, b)
	c.Outbound = append(c.Outbound, cp)
	return nil
}

func (c *LoopbackDatagramCarrier) RecvDatagram() ([]byte, error) {
	if len(c.Inbound) == 0 {
		return nil, nil
	}
	f := c.Inbound[0]
	c.Inbound = c.Inbound[1:]
	return f, nil
}

File: webapp/src/transport/csil/carrier.ts
// Carrier seams — the bring-your-own-carrier boundary (conventions doc §7).
//
// The library owns envelope codecs, framing, and lifecycle; the byte/datagram
// transport (the *carrier*) is injected. A host supplies a WebSocket, a
// WebTransport stream, a WebRTC unreliable DataChannel, QUIC datagrams, raw UDP,
// or anything else by implementing one of these interfaces — WITHOUT changing the
// library. That is the whole point of the seam, and it is why the conformance
// vectors (which test the codecs above the seam) need no real socket.
//
// HOW A HOST PLUGS IN A REAL CARRIER
// ----------------------------------
// The interfaces are deliberately tolerant of sync *or* async returns, so the
// in-memory loopbacks below stay synchronous for the codec tests while a real
// browser carrier returns Promises. RpcClient/RpcServer `await` every call, so
// both work unchanged.
//
// Browser WebSocket (CSIL-RPC / CSIL-Events — one envelope per binary frame):
//
//   class WebSocketCarrier implements FrameCarrier {
//     private inbox: Uint8Array[] = [];
//     private waiters: ((f: Uint8Array | null) => void)[] = [];
//     constructor(private ws: WebSocket) {
//       ws.binaryType = "arraybuffer";
//       ws.onmessage = (e) => this.deliver(new Uint8Array(e.data as ArrayBuffer));
//       ws.onclose = () => this.deliver(null);
//     }
//     private deliver(f: Uint8Array | null) {
//       const w = this.waiters.shift();
//       if (w) w(f); else if (f) this.inbox.push(f);
//     }
//     sendFrame(b: Uint8Array) { this.ws.send(b); }           // WS frames its own length
//     recvFrame(): Promise<Uint8Array | null> {
//       const f = this.inbox.shift();
//       return f ? Promise.resolve(f) : new Promise((res) => this.waiters.push(res));
//     }
//   }
//
// Note WebSocket already delimits messages, so no length prefix is added there.
// On a raw byte stream (TCP/TLS/Unix socket / a WebTransport bidi stream) the
// host instead uses `frameLengthPrefixed` on send and a `LengthPrefixedDeframer`
// fed by inbound chunks on receive — see those helpers below.
//
// WebRTC unreliable DataChannel / WebTransport datagrams (CSIL-Datagrams):
//
//   const dc = pc.createDataChannel("voice", { ordered: false, maxRetransmits: 0 });
//   class DataChannelCarrier implements DatagramCarrier {
//     // identical inbox/waiter pattern; dc.onmessage delivers one datagram each.
//     sendDatagram(b: Uint8Array) { dc.send(b); }
//     recvDatagram() { ... }
//   }
//
// In every case the library code is untouched; only this thin adapter changes.

import { MAX_FRAME_DEFAULT, TransportError, validateMaxFrame } from "./conventions.ts";

// A stream/frame carrier: sends and receives one *delimited message* at a time.
// Used by CSIL-RPC and CSIL-Events. `recvFrame` resolves to `null` at a clean end
// of stream. Implementations may be synchronous (the loopback) or asynchronous (a
// real WebSocket); consumers always `await`.
export interface FrameCarrier {
  sendFrame(bytes: Uint8Array): void | Promise<void>;
  recvFrame(): Uint8Array | null | Promise<Uint8Array | null>;
}

// A datagram carrier: sends and receives one self-contained datagram (each within
// the channel MTU), with no delivery or ordering guarantee. Used by CSIL-Datagrams.
export interface DatagramCarrier {
  sendDatagram(bytes: Uint8Array): void | Promise<void>;
  recvDatagram(): Uint8Array | null | Promise<Uint8Array | null>;
}

// Prefix `bytes` with a 4-byte big-endian length (CSIL stream framing), enforcing
// the max-frame guard before producing the frame.
export function frameLengthPrefixed(bytes: Uint8Array, max: number = MAX_FRAME_DEFAULT): Uint8Array {
  validateMaxFrame(max);
  if (bytes.length > max) throw TransportError.frameTooLarge(bytes.length, max);
  const out = new Uint8Array(4 + bytes.length);
  const view = new DataView(out.buffer);
  view.setUint32(0, bytes.length, false); // big-endian
  out.set(bytes, 4);
  return out;
}

// An incremental deframer for the 4-byte big-endian length-prefix wire. A host
// wiring a byte stream (TCP/TLS/WebTransport) feeds inbound chunks via `push` and
// drains complete frames via `next`. The max-frame guard rejects a hostile length
// prefix BEFORE buffering its claimed body, so a corrupt prefix cannot exhaust
// memory. This is the TS twin of the Rust `read_length_prefixed`.
export class LengthPrefixedDeframer {
  private buf: Uint8Array = new Uint8Array(0);
  private readonly max: number;

  constructor(max: number = MAX_FRAME_DEFAULT) {
    // Validated here rather than at the first frame, so a misconfigured deframer
    // is a construction-time error the host can surface at startup.
    this.max = validateMaxFrame(max);
  }

  // The limit this deframer enforces on an inbound length prefix.
  get maxFrame(): number {
    return this.max;
  }

  push(chunk: Uint8Array): void {
    const merged = new Uint8Array(this.buf.length + chunk.length);
    merged.set(this.buf, 0);
    merged.set(chunk, this.buf.length);
    this.buf = merged;
  }

  // Return the next complete frame, or `null` if not enough bytes have arrived.
  next(): Uint8Array | null {
    if (this.buf.length < 4) return null;
    const len = new DataView(this.buf.buffer, this.buf.byteOffset, 4).getUint32(0, false);
    if (len > this.max) throw TransportError.frameTooLarge(len, this.max);
    if (this.buf.length < 4 + len) return null;
    const frame = this.buf.slice(4, 4 + len);
    this.buf = this.buf.slice(4 + len);
    return frame;
  }
}

// An in-memory FrameCarrier backed by frame queues — for tests and for driving the
// codec without a socket. `outbound` collects what was sent; `inbound` feeds recv.
export class LoopbackFrameCarrier implements FrameCarrier {
  readonly outbound: Uint8Array[] = [];
  readonly inbound: Uint8Array[] = [];

  pushInbound(bytes: Uint8Array): void {
    this.inbound.push(bytes);
  }

  takeOutbound(): Uint8Array | undefined {
    return this.outbound.shift();
  }

  sendFrame(bytes: Uint8Array): void {
    this.outbound.push(bytes);
  }

  recvFrame(): Uint8Array | null {
    return this.inbound.shift() ?? null;
  }
}

// An in-memory DatagramCarrier — for tests and codec drives.
export class LoopbackDatagramCarrier implements DatagramCarrier {
  readonly outbound: Uint8Array[] = [];
  readonly inbound: Uint8Array[] = [];

  pushInbound(bytes: Uint8Array): void {
    this.inbound.push(bytes);
  }

  takeOutbound(): Uint8Array | undefined {
    return this.outbound.shift();
  }

  sendDatagram(bytes: Uint8Array): void {
    this.outbound.push(bytes);
  }

  recvDatagram(): Uint8Array | null {
    return this.inbound.shift() ?? null;
  }
}
